### PR TITLE
use safe accumulation for norm

### LINKED ISSUE #14126: Safe accumulation type for nd.norm
For FP16, reduction operators tend to loss precision if the accumulation data type remains fp16. Instead, the accumulation dtype should be in fp32. i.e. 
```
dtype = fp16 -> acc_type = fp32
dtype = fp32 -> acc_type = fp64
dtype = fp64 -> acc_type = fp64
```
We should do it for the norm op. 

Reference impl for softmax: https://github.com/apache/incubator-mxnet/pull/14098

### PR DESCRIPTION
**Pull request by szha**
## Description ##
use safe accumulation for norm. resolves #14126 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] use safe accumulation for norm

### CODE REVIEW

**Comment by roywei** [comment]
@mxnet-label-bot add[Backend, Operator, pr-awaiting-review]

**Comment by ptrendx** [comment]
I did not see it immediately from the code - what is the type of the result? Norm should output fp32 as it's result since it is very easy to overflow otherwise. Mean on the other hand can be done with just fp32 accumulator without changing the output type.

**Comment by szha** [comment]
I didn't change the output type yet. Looks like I should add a dtype option to norm too.

File: src/operator/tensor/broadcast_reduce-inl.cuh [hunk 1]
```diff
@@ -610,14 +609,22 @@ void ReduceImpl(cudaStream_t stream, const TBlob& small, const TBlob& lhs, const
...
+template<typename Reducer, int ndim, typename DType, typename OP, bool safe_acc = false>
 void Reduce(Stream<gpu> *s, const TBlob& small, const OpReqType req,
             const Tensor<gpu, 1, char>& workspace, const TBlob& big) {
   if (req == kNullOp) return;
   cudaStream_t stream = Stream<gpu>::GetStream(s);
   ReduceImplConfig<ndim> config =
     ConfigureReduceImpl<ndim, DType>(small.shape_, big.shape_, NULL, NULL);
-  ReduceImpl<Reducer, ndim, DType, OP>(stream, small, req, big, workspace, config);
+  if (safe_acc) {
+    MXNET_REAL_ACC_TYPE_SWITCH(mshadow::DataType<DType>::kFlag, DataType, AType, {
```
Comment (eric-haibin-lin) [inline src/operator/tensor/broadcast_reduce-inl.cuh]: Need to support acc type for int, etc

File: src/operator/mxnet_op.h [hunk 2]
```diff
@@ -273,25 +273,42 @@ inline int get_num_threads<cpu>(const int N) {
...
+      LOG(FATAL) << "This operation only support "         \
+                    "floating point types not uint8";      \
+    }                                                      \
     break;                                                 \
   case mshadow::kInt8:                                     \
-    LOG(FATAL) << "This operation only support "           \
-                  "floating point types not int8";         \
+    {                                                      \
+      typedef int8_t DType;                                \
+      typedef int8_t AType;                                \
```
Comment (eric-haibin-lin) [inline src/operator/mxnet_op.h]: can we support acc in int types, too?

**Pull request closed without merging**